src/backend/gl/gl_common.c

```diff
@@ -1125,16 +1288,222 @@ void *gl_create_blur_context(backend_t *base, enum blur_method method, void *arg
 		free(kernels);
 	}
 
+	free(extension);
+	// Restore LC_NUMERIC
+	setlocale(LC_NUMERIC, lc_numeric_old);
+	free(lc_numeric_old);
+
+	return success;
+}
+
+bool gl_create_dual_kawase_blur_context(void *blur_context, enum blur_method method,
+                                        void *args) {
+	bool success;
+	struct gl_blur_context *ctx = blur_context;
+
+	ctx->method = method;
+
+	auto blur_params = generate_dual_kawase_params(args);
+
+	// Specify required textures and FBOs
+	ctx->blur_texture_count = blur_params->iterations;
+	ctx->blur_fbo_count = blur_params->iterations;
+
+	ctx->resize_width += blur_params->expand;
+	ctx->resize_height += blur_params->expand;
+
+	ctx->npasses = 2;
+	ctx->blur_shader = ccalloc(ctx->npasses, gl_blur_shader_t);
+
+	char *lc_numeric_old = strdup(setlocale(LC_NUMERIC, NULL));
+	// Enforce LC_NUMERIC locale "C" here to make sure decimal point is sane
+	// Thanks to hiciu for reporting.
+	setlocale(LC_NUMERIC, "C");
+
+	// Dual-kawase downsample shader / program
+	auto down_pass = ctx->blur_shader;
+	{
+		// clang-format off
+		static const char *FRAG_SHADER_DOWN = GLSL(330,
+			uniform sampler2D tex_src;
+			uniform vec2 texture_size;
+			uniform float scale = 1.0;
+			uniform vec2 halfpixel;
+			in vec2 texcoord;
+			out vec4 out_color;
+			void main() {
+				float offset = %.7g;
+				vec2 uv = (texcoord * 2) / (texture_size * scale);
```

Comment:
Since the VAOs for the window quads supply the texture coordinates in actual pixel dimensions, we have to transform them back to normalized texture coordinates. Pixel coordinates are actually useful for the simplistic kernel blur with `texelFetch()`. The kawase kernels however relies on the linear filtering of `texture2D()` which works with normalized coordinates.
Could we benefit from supplying normalized texture coords in `x_rect_to_coords()` with a flag in case we want to blur with dual-kawase? We should be able to skip updating the `texture_size` uniform with every iteration. (And `scale` would not be needed anymore in the fragment shaders.)

Reply:
`texelFetch()` does not implement any repeat/clamp logic like `texture2D()` does. This would have to be manually implemented in the shader requiring the actual source texture dimensions. Currently this leads to darkened blur around screen borders in the kernel blur implementation.

I'd suggest adopting the kernel blur shader to use `texture2D()` as well with correctly setup textures and provide normalized coordinates in `x_rect_to_coords()` when requested.